packages/xstate-store/src/atom.ts

```diff
@@ -110,41 +151,23 @@ export function createAtom<T>(
           link(atom, activeSub);
         }
         return atom._snapshot;
-      },
-      _update(): boolean {
-        const prevSub = activeSub;
-        const compare = options?.compare ?? Object.is;
-        activeSub = atom as InternalReadonlyAtom<T>;
-        startTracking(atom as InternalReadonlyAtom<T>);
-        try {
-          const oldValue = atom._snapshot;
-          const read = (atom: Readable<any>) => atom.get();
-          const newValue = getter(read);
-          if (oldValue === undefined || !compare(oldValue, newValue)) {
-            atom._snapshot = newValue;
-            return true;
-          }
-          return false;
-        } finally {
-          activeSub = prevSub;
-          endTracking(atom as InternalReadonlyAtom<T>);
-        }
       }
     });
   } else {
     Object.assign<BaseAtom<T>, Pick<Atom<T>, 'set'>>(atom, {
       set(valueOrFn: T | ((prev: T) => T)): void {
-        const compare = options?.compare ?? Object.is;
-        const value =
-          typeof valueOrFn === 'function'
-            ? (valueOrFn as (prev: T) => T)(atom._snapshot)
-            : valueOrFn;
-        if (compare(atom._snapshot, value)) return;
-        atom._snapshot = value;
-        const { _subs: subs } = atom;
-        if (subs !== undefined) {
-          propagate(subs);
-          processEffectNotifications();
+        if (
+          atom._update(
+            typeof valueOrFn === 'function'
+              ? () => (valueOrFn as (prev: T) => T)(atom._snapshot)
+              : () => valueOrFn
+          )
```

Comment:
it would be better to push the function resolving to the `atom._update` itself, this else branch here introduced a redundant function wrapper for the value, one that is going to be called synchronously *anyway*

Reply:
[c54de6b](https://github.com/statelyai/xstate/pull/5266/commits/c54de6bde210c577c414b9d08f3ce78fd47b22bc)